bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/ParentResourceValueMapValueInjector.java

```diff
@@ -0,0 +1,85 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.adobe.acs.commons.models.injectors.impl;
+
+import com.adobe.acs.commons.models.injectors.annotation.ParentResourceValueMapValue;
+import com.day.cq.wcm.api.NameConstants;
+import com.day.cq.wcm.foundation.forms.MergedValueMap;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.models.spi.DisposalCallbackRegistry;
+import org.apache.sling.models.spi.Injector;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Injects the parent component property
+ */
+@Component(property = Constants.SERVICE_RANKING + ":Integer=" + Integer.MAX_VALUE, service = Injector.class)
+public class ParentResourceValueMapValueInjector implements Injector {
+
+    @Override
+    public String getName() {
+        return ParentResourceValueMapValue.SOURCE;
+    }
+
+    @Override
+    public Object getValue(Object adaptable, String name, Type type, AnnotatedElement annotatedElement, DisposalCallbackRegistry disposalCallbackRegistry) {
+        Integer maxLevel = annotatedElement.getAnnotation(ParentResourceValueMapValue.class).maxLevel();
+        if (adaptable instanceof Resource) {
+            return getInheritedProperty((Resource)adaptable, name, type, maxLevel);
+
+        } else if (adaptable instanceof SlingHttpServletRequest) {
+            Resource adaptableResource = ((SlingHttpServletRequest) adaptable).getResource();
+            return getInheritedProperty(adaptableResource, name, type, maxLevel);
+        }
+        return null;
+    }
+
+    /**
+     * Returns the property by iterating through the parent component resources excluding the cq:Page resource
+     * @param resource Resource
+     * @param propertyName Property name to be retrieved
+     * @param declaredType Declared class type
+     * @param maxLevel Max level of parents to iterate
+     * @return Object
+     */
+    private Object getInheritedProperty(Resource resource, String propertyName, Type declaredType, Integer maxLevel) {
+        List<Resource> parentResources = new ArrayList<>();
+        int count = 0;
+        Resource parentResource = resource.getParent();
+        do {
+            if (parentResource != null && !parentResource.getResourceType().equals(NameConstants.NT_PAGE)) {
+                parentResources.add(resource.getParent());
+                count++;
+            } else {
+                break;
+            }
+            parentResource = parentResource.getParent();
+        }
+        while (count < maxLevel || maxLevel <= 0);
+        return new MergedValueMap(parentResources).get(propertyName, (Class<?>) declaredType);
```

Comment:
Getting all of the parent valuemaps and merging them all just to get a single property feels a bit heavy.  Within the loop couldn't we simply call `resource.getValueMap().get(propertyName, declaredType)` and then `break;` if a non-null value is found?  That way we dont even look at higher resources in the tree than necessary when the property is at the lower levels, and we dont have to bother merging a bunch of maps.

Thoughts?

Reply:
Yeah, this makes sense if the max-level value is high and has to iterate through a bunch of nodes or if each node has many properties.